File: backend/ts_template/devdonalds.ts
// import express, { Request, Response } from "express";
import express from "express"; // runtime import
import type { Request, Response } from "express"; // types only


// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

interface RecipeItem {
  name: string;
  quantity: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());
export default app;

// Store your recipes here!
const cookbook: any = null;
let ingredient: Map<String, number> = new Map();
let recipe: Map<String, String> = new Map();

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  // TODO: implement me
  let retString: string = ""
  let start: boolean = true

  for (const char of recipeName) {
    const Alpha = /^[a-zA-Z]$/.test(char)
    if (Alpha) {
      if (start) {
        retString += char.toUpperCase()
        start = false
      } else {
        retString += char.toLocaleLowerCase()
      }
    } else if (char === "-" || char === "_" || char === " ") {
      if (!start) { 
        retString += " "
      }
      start = true
    }

  }
  if (retString.length === 0 ) {
    return null
  }
  return retString
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const entry = req.body
  let type = entry.type 
  let name = entry.name 
  console.log(name)
  
  if (ingredient.has(name) || recipe.has(name)) {
    res.status(400).send("entry already exists")
    return
  }

  if (type === "recipe") {
    let requirements = entry.requiredItems
    for (let items of requirements) {
      if (items.length > 2) {
        res.status(400).send("too many ingredient parameters")
        return
      }
    }

    recipe.set(name, requirements)

  } else if (type == "ingredient") {
     let time = entry.cookTime
     if (time < 0) {
      res.status(400).send("invalid cooktime")
      return
     }
     if (time === 0) {
      time = -1
     }
     ingredient.set(name, time)

  } else {
    res.status(400).send("invalid type")
    return
  }

  res.status(200).send("added onto the database!")
});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  let name: string = req.query.name
  let totalTime: number = 0;

  let items: Map<string, number> = new Map()
  let recipes = []

  if (!recipe.has(name)) {
    res.status(400).send(`hello: ${name}`)
    return
  }

  console.log(recipe.get(name))
  for (let item of recipe.get(name) as unknown as RecipeItem[]) {
    if (recipe.has(item.name)) {
      recipes.push([item.name, item.quantity])
    }
    else {
      items.set(item.name, item.quantity)
    }
  }
  
  while (recipes.length != 0) {
    let something = recipes[0];

    let ing = something[0];
    let amount = something[1];
    for (let item of recipe.get(ing)) {
      let json = JSON.parse(item)
      if (recipe.has(json.name)) {
        recipes.push([json.name, amount * json.quantity])
      }
      else {
        items.set(json.name, json.quantity * amount)
      }
    }
  }
  
  let array = []

  for (const keys of Array.from(items.keys())) {
    if (! ingredient.has(keys)) {
      res.status(400).send(`${keys} does not exist`)
      return
    }
    if (ingredient.get(keys) === -1) {
      ingredient.set(keys, 0) 
    }
    totalTime += items.get(keys) * ingredient.get(keys)
    array.push({"name": keys, "quantity": items.get(keys)})
  }

  let output = {
		"name": name,
		"cookTime": totalTime,
		"ingredients": array
	}

  res.status(200).send(output)

});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
#!/usr/bin/env python3

from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# added libraries
from collections import defaultdict
import json
import unittest

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int

 
# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = None
recipe = defaultdict(str)
ingredient = defaultdict(str)

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that is legible as per spec
def parse_handwriting(recipeName: str) -> Union[str , None]:
	# TODO: implement me
	finalString = ""
	first = True

	for character in recipeName:
		if character.isalpha():
			if first:
				finalString += character.upper()
			else:
				finalString += character.lower()
			first = False

		elif character == '-' or character == '_' or character == ' ':
			if not first:
				finalString += " "
			first = True

	return finalString if len(finalString) > 0 else None


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():

	entry = request.get_json()

	type = entry["type"]
	name = entry["name"]
	if recipe[name] or ingredient[name]:
		return 'entry already exists', 400

	if type == 'recipe':
		requirements = entry["requiredItems"]
		for items in requirements:
			if len(items) > 2:
				return 'too many elements within requiredItems', 400
		
		recipe[name] = requirements

	elif type == 'ingredient':
		time = entry['cookTime'] 
		if time < 0:
			return 'invalid cooking time', 400

		if time == 0:
			time = -1
		ingredient[name] = time

	else:
		return 'invalid type', 400


	return f'added {name} onto the database', 200



# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	name = request.args.get('name')
	totalTime = 0
	items = defaultdict(int)
	recipes = []

	if not recipe[name]:
		return 'not a proper reciepe', 400

	for item in recipe[name]:
		if recipe[item['name']]:
			recipes.append([item['name'], item['quantity']])
		else:
			items[item['name']] += item['quantity']
	
	
	while recipes:
		new, amount = recipes.pop(0)

		for item in recipe[new]:
			if recipe[item['name']]:
				recipes.append([item['name'], amount * item['quantity']])
			else:
				items[item['name']] += (item['quantity'] * amount)
	array = []

	output = []
	for keys in items.keys():
		if not ingredient[keys]:
			return f'{keys} does not exist', 400
		if ingredient[keys] == -1:
			ingredient[keys] = 0
		totalTime += items[keys] * ingredient[keys]
		output.append([items[keys] , ingredient[keys]])
		array.append({"name": f"{keys}", "quantity": f'{items[keys]}'})


	output = {
		"name": f"{name}",
		"cookTime": f"{totalTime}",
		"ingredients": f"{array}"
	}

	return output, 200


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
